manifests/init.pp

```diff
@@ -180,12 +192,22 @@
 ) inherits katello::params {
 
   include katello::repo
-  include katello::candlepin
-  include katello::qpid
-  include katello::pulp
-  Class['katello::repo'] -> Class['katello::pulp']
-  include katello::application
-  Class['katello::repo'] -> Class['katello::application']
-  Class['katello::qpid'] -> Class['katello::candlepin'] -> Class['katello::application']
-
+  if $manage_candlepin {
+    include katello::candlepin
+    Class['katello::qpid'] -> Class['katello::candlepin']
```

Comment:
Does this work if the actual inclusion is later?

Reply:
I'm just wondering why we need the relationship definitions in the first place?
By default Puppet applies resources in the order they are declared in the manifest. Then there would be no need to account for every combination - we just need to make sure the ordering of includes is correct in the first place.
Is there something special here that I'm missing?

Reply:
> By default Puppet applies resources in the order they are declared in the manifest.

This is in no way a guarantee. However, Candlepin will fail to start if qpid is not available and configured. More specifically: it needs the event queue to be defined which we do in `katello::qpid`.

Part of this is old school Puppet. Prior to Puppet 4 the order was actually randomized and you had to be explicit. However, this particular case is actually explicitly a requirement and the app is known to fail hard if the condition is not met.

Reply:
Even though it should be no problem to declare relationships between resources before declaring the resources itself, I did change the ordering of candlepin and qpid to make the order consistent with the relations.

> Relationships are not limited by evaluation-order; you can declare a relationship with a resource before that resource has been declared.
(from https://puppet.com/docs/puppet/6.10/lang_relationships.html)